compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java

```diff
@@ -1739,6 +1739,13 @@ public void setForeachTypedBindingPatternType(BLangForeach foreachNode) {
                     case TypeTags.NEVER:
                         varType = symTable.neverType;
                         break;
+                    case TypeTags.ARRAY:
```

Comment:
How can the constraint be an array?

Reply:
Its happen when `xml<xml:Element[]> element = xml ``;`

Reply:
But this is invalid, yeah? So we don't have to handle foreach for it.

Reply:
yes. but if we didn't handle it. It will throw a runtime error.

Reply:
https://github.com/ballerina-platform/ballerina-lang/blob/03275ec1c2e80c558adeb78165eaf534519b40c2/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/Types.java#L1744

If we didn't handle it, It will go to the default case. Then it will throw a class cast exception.
Because `BArrayType` can't convert into a `BUnionType`.

Reply:
Then it will still fail for things like

```ballerina
public function main() {
    xml<[int, string]> x = xml ``;

    foreach var item in x {
        
    }
}
```

So, IMO the fix should be to add a case for UNION instead of considering it to be the default, and make the default case do something like L1818-L1822.

Reply:
Yes, That will be the better option. I will update that.

Reply:
```
public function main() {
    xml<xmL:Element[] & readonly> x = xml ``;

    foreach var item in x {
        
    }
}
```
There are two possible error messages we can display to user.
* incompatible types: 'xml' cannot be constrained with '(xml:Element[] & readonly)' 
* incompatible types: 'xml<xml:Element[]>' is not an iterable collection

IMO, first one is better because ballerina didn't support `xml:Element[] & readonly`.
WDYT?

Reply:
Don't these correspond to the two different statements? The first error we must definitely log. The second one is not required IMO, and is only noise since it will go away when the user fixes the first issue anyway.

Reply:
Yes. I will add first message.

Reply:
Updated - 92b1e95bcd91c8dc89775d0f23f3a91ef0236f75